fits-schema is no longer accessible

Because of https://github.com/spacetelescope/asdf/pull/654, our `datamodels` schemas no longer validate. @drdavella 

The error message:
```
$ pytest jwst/datamodels/schemas/core.schema.yaml
============================= test session starts ==============================
platform darwin -- Python 3.6.8, pytest-4.3.0, py-1.8.0, pluggy-0.9.0
rootdir: /Users/jdavies/dev/jwst, inifile: setup.cfg
plugins: xdist-1.26.1, remotedata-0.3.1, openfiles-0.3.2, forked-1.0.2, doctestplus-0.2.0, arraydiff-0.3, ci-watson-0.3
collected 1 item

jwst/datamodels/schemas/core.schema.yaml F                               [100%]

=================================== FAILURES ===================================
_______________________________________  _______________________________________

cls = <class '_pytest.runner.CallInfo'>
func = <function call_runtest_hook.<locals>.<lambda> at 0xb19206378>
when = 'call'
reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)

    @classmethod
    def from_call(cls, func, when, reraise=None):
        #: context of invocation: one of "setup", "call",
        #: "teardown", "memocollect"
        start = time()
        excinfo = None
        try:
>           result = func()

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/_pytest/runner.py:226: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>       lambda: ihook(item=item, **kwds), when=when, reraise=reraise
    )

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/_pytest/runner.py:198: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtest_call'>, args = ()
kwargs = {'item': <AsdfSchemaItem /Users/jdavies/dev/jwst/jwst/datamodels/schemas/core.schema.yaml>}
notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.spec and self.spec.argnames:
            notincall = (
                set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())
            )
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call".format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self.get_hookimpls(), kwargs)

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/pluggy/hooks.py:289: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x1069b8f28>
hook = <_HookCaller 'pytest_runtest_call'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/Users/jdavies/miniconda3/envs/jwst_dev/lib/pyt...t 0x10b9e23c8>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0xb1925de48>>]
kwargs = {'item': <AsdfSchemaItem /Users/jdavies/dev/jwst/jwst/datamodels/schemas/core.schema.yaml>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/pluggy/manager.py:68: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtest_call'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/Users/jdavies/miniconda3/envs/jwst_dev/lib/pyt...t 0x10b9e23c8>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0xb1925de48>>]
kwargs = {'item': <AsdfSchemaItem /Users/jdavies/dev/jwst/jwst/datamodels/schemas/core.schema.yaml>}

    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
        methods,
        kwargs,
>       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
    )

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/pluggy/manager.py:62: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook_impls = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/Users/jdavies/miniconda3/envs/jwst_dev/lib/pyt...t 0x10b9e23c8>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0xb1925de48>>]
caller_kwargs = {'item': <AsdfSchemaItem /Users/jdavies/dev/jwst/jwst/datamodels/schemas/core.schema.yaml>}
firstresult = False

    def _multicall(hook_impls, caller_kwargs, firstresult=False):
        """Execute a call into multiple python functions/methods and return the
        result(s).
    
        ``caller_kwargs`` comes from _HookCaller.__call__().
        """
        __tracebackhide__ = True
        results = []
        excinfo = None
        try:  # run impl and wrapper setup functions in a loop
            teardowns = []
            try:
                for hook_impl in reversed(hook_impls):
                    try:
                        args = [caller_kwargs[argname] for argname in hook_impl.argnames]
                    except KeyError:
                        for argname in hook_impl.argnames:
                            if argname not in caller_kwargs:
                                raise HookCallError(
                                    "hook call must provide argument %r" % (argname,)
                                )
    
                    if hook_impl.hookwrapper:
                        try:
                            gen = hook_impl.function(*args)
                            next(gen)  # first yield
                            teardowns.append(gen)
                        except StopIteration:
                            _raise_wrapfail(gen, "did not yield")
                    else:
                        res = hook_impl.function(*args)
                        if res is not None:
                            results.append(res)
                            if firstresult:  # halt further impl calls
                                break
            except BaseException:
                excinfo = sys.exc_info()
        finally:
            if firstresult:  # first result hooks return a single value
                outcome = _Result(results[0] if results else None, excinfo)
            else:
                outcome = _Result(results, excinfo)
    
            # run all wrapper post-yield blocks
            for gen in reversed(teardowns):
                try:
                    gen.send(outcome)
                    _raise_wrapfail(gen, "has second yield")
                except StopIteration:
                    pass
    
>           return outcome.get_result()

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/pluggy/callers.py:208: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pluggy.callers._Result object at 0xb195ccbe0>

    def get_result(self):
        """Get the result(s) for this hook call.
    
        If the hook was marked as a ``firstresult`` only a single value
        will be returned otherwise a list of results.
        """
        __tracebackhide__ = True
        if self._excinfo is None:
            return self._result
        else:
            ex = self._excinfo
            if _py3:
>               raise ex[1].with_traceback(ex[2])

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/pluggy/callers.py:80: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook_impls = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/Users/jdavies/miniconda3/envs/jwst_dev/lib/pyt...t 0x10b9e23c8>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0xb1925de48>>]
caller_kwargs = {'item': <AsdfSchemaItem /Users/jdavies/dev/jwst/jwst/datamodels/schemas/core.schema.yaml>}
firstresult = False

    def _multicall(hook_impls, caller_kwargs, firstresult=False):
        """Execute a call into multiple python functions/methods and return the
        result(s).
    
        ``caller_kwargs`` comes from _HookCaller.__call__().
        """
        __tracebackhide__ = True
        results = []
        excinfo = None
        try:  # run impl and wrapper setup functions in a loop
            teardowns = []
            try:
                for hook_impl in reversed(hook_impls):
                    try:
                        args = [caller_kwargs[argname] for argname in hook_impl.argnames]
                    except KeyError:
                        for argname in hook_impl.argnames:
                            if argname not in caller_kwargs:
                                raise HookCallError(
                                    "hook call must provide argument %r" % (argname,)
                                )
    
                    if hook_impl.hookwrapper:
                        try:
                            gen = hook_impl.function(*args)
                            next(gen)  # first yield
                            teardowns.append(gen)
                        except StopIteration:
                            _raise_wrapfail(gen, "did not yield")
                    else:
>                       res = hook_impl.function(*args)

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/pluggy/callers.py:187: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <AsdfSchemaItem /Users/jdavies/dev/jwst/jwst/datamodels/schemas/core.schema.yaml>

    def pytest_runtest_call(item):
        _update_current_test_var(item, "call")
        sys.last_type, sys.last_value, sys.last_traceback = (None, None, None)
        try:
>           item.runtest()

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/_pytest/runner.py:123: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <AsdfSchemaItem /Users/jdavies/dev/jwst/jwst/datamodels/schemas/core.schema.yaml>

    def runtest(self):
        # Make sure that each schema itself is valid.
        schema_tree = schema.load_schema(
            self.schema_path, resolver=_resolver, resolve_references=True)
>       schema.check_schema(schema_tree)

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/tests/schema_tester.py:103: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

schema = OrderedDict([('type', 'object'), ('properties', OrderedDict([('meta', OrderedDict([('type', 'object'), ('properties', ...ng'), ('fits_keyword', 'MASTERBG')]))]))]))]))]))])), ('$schema', 'http://stsci.edu/schemas/fits-schema/fits-schema')])

    def check_schema(schema):
        """
        Check a given schema to make sure it is valid YAML schema.
        """
        # We also want to validate the "default" values in the schema
        # against the schema itself.  jsonschema as a library doesn't do
        # this on its own.
    
        def validate_default(validator, default, instance, schema):
            if not validator.is_type(instance, 'object'):
                return
    
            if 'default' in instance:
                with instance_validator.resolver.in_scope(scope):
                    for err in instance_validator.iter_errors(
                            instance['default'], instance):
                        yield err
    
        VALIDATORS = util.HashableDict(
            mvalidators.Draft4Validator.VALIDATORS.copy())
        VALIDATORS.update({
            'default': validate_default
        })
    
        meta_schema_id = schema.get('$schema', YAML_SCHEMA_METASCHEMA_ID)
>       meta_schema = load_schema(meta_schema_id, mresolver.default_resolver)

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/schema.py:586: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

url = 'http://stsci.edu/schemas/fits-schema/fits-schema'
resolver = <function default_resolver at 0x1165a2f28>
resolve_references = False, resolve_local_refs = False

    @lru_cache()
    def load_schema(url, resolver=None, resolve_references=False,
                    resolve_local_refs=False):
        """
        Load a schema from the given URL.
    
        Parameters
        ----------
        url : str
            The path to the schema
    
        resolver : callable, optional
            A callback function used to map URIs to other URIs.  The
            callable must take a string and return a string or `None`.
            This is useful, for example, when a remote resource has a
            mirror on the local filesystem that you wish to use.
    
        resolve_references : bool, optional
            If `True`, resolve all `$ref` references.
    
        resolve_local_refs : bool, optional
            If `True`, resolve all `$ref` references that refer to other objects
            within the same schema. This will automatically be handled when passing
            `resolve_references=True`, but it may be desirable in some cases to
            control local reference resolution separately.
        """
        if resolver is None:
            resolver = mresolver.default_resolver
        loader = _make_schema_loader(resolver)
        if url in HARDCODED_SCHEMA:
            schema = HARDCODED_SCHEMA[url]()
        else:
>           schema, url = loader(url)

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/schema.py:367: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

url = 'file:///Users/jdavies/miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/schemas/stsci.edu/fits-schema/fits-schema.yaml'

    def load_schema(url):
        url = resolver(url)
>       return _load_schema(url)

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/schema.py:281: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

url = 'file:///Users/jdavies/miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/schemas/stsci.edu/fits-schema/fits-schema.yaml'

    @lru_cache()
    def _load_schema(url):
>       with generic_io.get_file(url) as fd:

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/schema.py:269: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

init = 'file:///Users/jdavies/miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/schemas/stsci.edu/fits-schema/fits-schema.yaml'
mode = 'r', uri = None, close = False

    def get_file(init, mode='r', uri=None, close=False):
        """
        Returns a `GenericFile` instance suitable for wrapping the given
        object `init`.
    
        If passed an already open file-like object, it must be opened for
        reading/writing in binary mode.  It is the caller's responsibility
        to close it.
    
        Parameters
        ----------
        init : object
            `init` may be:
    
            - A `bytes` or `unicode` file path or ``file:`` or ``http:``
              url.
    
            - A Python 2 `file` object.
    
            - An `io.IOBase` object (the default file object on Python 3).
    
            - A ducktyped object that looks like a file object.  If `mode`
              is ``"r"``, it must have a ``read`` method.  If `mode` is
              ``"w"``, it must have a ``write`` method.  If `mode` is
              ``"rw"`` it must have the ``read``, ``write``, ``tell`` and
              ``seek`` methods.
    
            - A `GenericFile` instance, in which case it is wrapped in a
              `GenericWrapper` instance, so that the file is closed when
              only when the final layer is unwrapped.
    
        mode : str
            Must be one of ``"r"``, ``"w"`` or ``"rw"``.
    
        uri : str
            Sets the base URI of the file object.  This will be used to
            resolve any relative URIs contained in the file.  This is
            redundant if `init` is a `bytes` or `unicode` object (since it
            will be the uri), and it may be determined automatically if
            `init` refers to a regular filesystem file.  It is not required
            if URI resolution is not used in the file.
    
        close : bool
            If ``True``, closes the underlying file handle when this object is
            closed. Defaults to ``False``.
    
        Returns
        -------
        fd : GenericFile
    
        Raises
        ------
        ValueError, TypeError, IOError
        """
        if mode not in ('r', 'w', 'rw'):
            raise ValueError("mode must be 'r', 'w' or 'rw'")
    
        if init in (sys.__stdout__, sys.__stdin__, sys.__stderr__):
            init = os.fdopen(init.fileno(), init.mode + 'b')
    
        if isinstance(init, (GenericFile, GenericWrapper)):
            if mode not in init.mode:
                raise ValueError(
                    "File is opened as '{0}', but '{1}' was requested".format(
                        init.mode, mode))
            return GenericWrapper(init)
    
        elif isinstance(init, (str, pathlib.Path)):
            parsed = urlparse.urlparse(str(init))
            if parsed.scheme in ['http', 'https']:
                if 'w' in mode:
                    raise ValueError(
                        "HTTP connections can not be opened for writing")
                return _make_http_connection(init, mode, uri=uri)
            elif parsed.scheme in _local_file_schemes:
                if mode == 'rw':
                    realmode = 'r+b'
                else:
                    realmode = mode + 'b'
                # Windows paths are not URIs, and so they should not be parsed as
                # such. Otherwise, the drive component of the path can get lost.
                # This is not an ideal solution, but we can't use pathlib here
                # because it doesn't handle URIs properly.
                if sys.platform.startswith('win') and parsed.scheme in string.ascii_letters:
                    realpath = str(init)
                else:
                    realpath = url2pathname(parsed.path)
                if mode == 'w':
                    fd = atomicfile.atomic_open(realpath, realmode)
                else:
>                   fd = open(realpath, realmode)
E                   FileNotFoundError: [Errno 2] No such file or directory: '/Users/jdavies/miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/schemas/stsci.edu/fits-schema/fits-schema.yaml'

../../miniconda3/envs/jwst_dev/lib/python3.6/site-packages/asdf/generic_io.py:1198: FileNotFoundError
=========================== 1 failed in 0.50 seconds ===========================
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,6 +1,13 @@
 0.13.2 (Unreleased)
 ===================
 
+datamodels
+----------
+
+- Fix ``url_mapper`` for fits-schema to allow URLs with of the format
+  http://stsci.edu/schemas/fits-schema/ to map to the correct location
+  in the ``jwst`` package. [#3239]
+
 0.13.1 (2019-03-07)
 ===================
 

--- a/jwst/datamodels/extension.py
+++ b/jwst/datamodels/extension.py
@@ -4,6 +4,8 @@ from asdf import util
 
 SCHEMA_PATH = os.path.abspath(
     os.path.join(os.path.dirname(__file__), 'schemas'))
+METASCHEMA_PATH = os.path.abspath(
+    os.path.join(os.path.dirname(__file__), 'metaschema'))
 
 URL_PREFIX = 'http://jwst.stsci.edu/schemas/'
 
@@ -26,5 +28,8 @@ class BaseExtension(AsdfExtension):
 
     @property
     def url_mapping(self):
-        return [(URL_PREFIX,
-                 util.filepath_to_url(SCHEMA_PATH) + '/{url_suffix}')]
+        return [
+            (URL_PREFIX, util.filepath_to_url(SCHEMA_PATH) + '/{url_suffix}'),
+            ('http://stsci.edu/schemas/fits-schema/',
+                util.filepath_to_url(METASCHEMA_PATH) + '/{url_suffix}.yaml'),
+        ]

--- a/jwst/datamodels/metaschema/fits-schema.yaml
+++ b/jwst/datamodels/metaschema/fits-schema.yaml
@@ -7,7 +7,7 @@ description: |
   to converting models to and from FITS files.
 
 allOf:
-  - $ref: "http://stsci.edu/schemas/asdf-schema/0.1.0/asdf-schema"
+  - $ref: "http://stsci.edu/schemas/asdf/asdf-schema-1.0.0"
   - type: object
     properties:
       fits_keyword:

